### ISSUE TITLE
[ENH] Transformed Target Forecaster (prediction quantiles & intervals) implementation request. 

### ISSUE THREAD

**Issue by HishamSalem** [alan-turing-institute/sktime#2767]
The Transformed Target forecaster is not implemented in sktime, is it possible if it gets implemented? Is there an alternative to manually implement prediction intervals ensemble models?

Have the basic forecasting coverage with the option to utilize bootstrap for confidence intervals  
https://otexts.com/fpp2/prediction-intervals.html

If there is a way to implement this for ensemble XGBoost models that would be excellent :)

**Additional context**

**Comment by fkiraly** [comment]
@HishamSalem, the `TransformedTargetForecaster` does implement prediction intervals currently! For that, the forecaster in it needs to support prediction intervals.

If you want to *add* prediction intervals to a forecaster that doesn´t natively have them, you can use one of the wrappers that add them to any forecaster, like `NaiveVariance` or `ConvormalIntervals`.

**Comment by HishamSalem** [comment]
@fkiraly  Thank you for the advice, I tried this and it still gives a not implemented error. 

Code:

from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveVariance
from sktime.forecasting.compose import TransformedTargetForecaster,make_reduction #Make reduction
from xgboost import XGBRegressor

y = load_airline()
forecaster = TransformedTargetForecaster(
    [        
      (
          "forecast",
          make_reduction(XGBRegressor()   
              ,
              scitype="tabular-regressor",
              window_length=16,
              strategy="recursive",),
      ),
  ])
#forecast
variance_forecaster = NaiveVariance(forecaster)
variance_forecaster.fit(y)
var_pred = variance_forecaster.predict_interval(fh=[1,2,3],coverage=0.9)



NotImplementedError: Generating in-sample predictions is not yet implemented for RecursiveTabularRegressionForecaster.

**Comment by HishamSalem** [comment]
I do not think the ConformalIntervals is in my modules yet. Do I need to update sktime? 

I did try to take the module directly into my pc, can you share the installation wheel so that I can import ConformalIntervals?

**Comment by fkiraly** [comment]
bootstrap bagging is in the works here, btw:
https://github.com/alan-turing-institute/sktime/pull/2248

**Comment by HishamSalem** [comment]
I updated the package as instructed and I do get an indexing error specifically for sliding window models. The error mainly has to do with the make reduction package. 

from sktime.datasets import load_airline
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.naive import NaiveForecaster
y = load_airline()
forecaster = TransformedTargetForecaster(
    [        
      (
          "forecast",
          make_reduction(XGBRegressor()   
              ,
              scitype="tabular-regressor",
              window_length=16,
              strategy="recursive",),
      ),
  ])
Variance=NaiveVariance(forecaster)
conformal_forecaster = ConformalIntervals(Variance)
conformal_forecaster.fit(y, fh=[1,2,3])
pred_int = conformal_forecaster.predict_interval()


[<ipython-input-33-5c283e44b7db>](https://localhost:8080/#) in <module>()
     16 Variance=NaiveVariance(forecaster)
     17 conformal_forecaster = ConformalIntervals(Variance)
---> 18 conformal_forecaster.fit(y, fh=[1,2,3])
     19 pred_int = conformal_forecaster.predict_interval()

8 frames
[/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py](https://localhost:8080/#) in _align_frame(self, indexer, df)
   2199             return val
   2200 
-> 2201         raise ValueError("Incompatible indexer with DataFrame")
   2202 
   2203 

ValueError: Incompatible indexer with DataFrame

**Comment by fkiraly** [comment]
@HishamSalem, content-wise, you should not be wrapping a forecaster in *two* of the interval generating wrappers - use at most one.

In theory, though, it should still work if you wrap them in both. @bethrice44, any idea what is happening?

**Comment by fkiraly** [comment]
This very similar code works (I used `RandomForest` instead of `XGBRegressor` to stay inside `sktime` dependency set)

```python

from sklearn.ensemble import RandomForestRegressor

from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.naive import NaiveForecaster
y = load_airline()

f = make_reduction(
    RandomForestRegressor(),
    scitype="tabular-regressor",
    window_length=16,
    strategy="recursive",
)

conformal_forecaster = ConformalIntervals(f)
conformal_forecaster.fit(y, fh=[1,2,3])
pred_int = conformal_forecaster.predict_interval()
```

**Comment by fkiraly** [comment]
Hm, the same code fails when wrapped in `TransformedTargetForecaster`:

```python
from sklearn.ensemble import RandomForestRegressor

from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction,TransformedTargetForecaster
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.naive import NaiveForecaster

y = load_airline()

forecaster = TransformedTargetForecaster(
    [(
        "forecast",
        make_reduction(
            RandomForestRegressor(),
            scitype="tabular-regressor",
            window_length=16,
            strategy="recursive",
        ),
    )]
)

conformal_forecaster = ConformalIntervals(forecaster)
conformal_forecaster.fit(y, fh=[1,2,3])
pred_int = conformal_forecaster.predict_interval()
```

**Comment by HishamSalem** [comment]
@fkiraly That makes sense, thanks for the clarification. I think problem is associated with pipeline might be the problem because it does work when using the just the sklearn model as is however when integrating make_pipeline as well the same error pops up.  

Sample:
from sklearn.ensemble import RandomForestRegressor

from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction,TransformedTargetForecaster
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.naive import NaiveForecaster

y = load_airline()

forecaster = make_pipeline(
    Deseasonalizer(),    
    make_reduction(RandomForestRegressor(), window_length=12, strategy="recursive"),
)

conformal_forecaster = ConformalIntervals(forecaster)
conformal_forecaster.fit(y, fh=[1,2,3])
pred_int = conformal_forecaster.predict_interval()

However when we avoid pipelineing

from sklearn.ensemble import RandomForestRegressor

from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction,TransformedTargetForecaster
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.naive import NaiveForecaster

y = load_airline()

forecaster = make_reduction(RandomForestRegressor(), window_length=52, strategy="recursive")

conformal_forecaster = ConformalIntervals(forecaster)
conformal_forecaster.fit(y, fh=[1,2,3])
pred_int = conformal_forecaster.predict_interval()

**Comment by fkiraly** [comment]
PS, @HishamSalem, you can pretty-print python code blocks by prefacing them with ` \`\`\`python ` and ending them with ` \`\`\` `

**Comment by fkiraly** [comment]
from the slack conversation with @bethrice44 

in the current implementation of ConformalIntervals it is not set to a specific type, but copies it over from y_inner_mtype, so can be anything dependent on what the internal forecaster wants
maybe not a good idea if you want it to be a specific type, always
might be a bug
the tags X_inner_mtype and y_inner_mtype need to be set to whatever _compute_sliding_residuals supports

so I guess pd.DataFrame or pd.Series
the copying of "inner" to "outer" happens in line 131

option 1 to fix this: set the y_inner_mtype tag to sth fixed
option 2: in compute_sliding_residuals, add a convert_to line at the start, and leave the "clone tags" line. That way, there is no superfluous conversion happening in fit

**Comment by fkiraly** [comment]
@HishamSalem, the bugfix should resolve both your recent and original problem, hopefully.
(on current developer version, `main` branch, and to be released with 0.12.1)
If not, kindly relopen.
